### DISCUSSION TITLE
Have a selector wait until the state has been initialised (or some other condition)

### DISCUSSION THREAD

**Discussion by jits** [ngxs/store#1756] [Ideas]
### Relevant Package

This feature request is for @ngxs/store

### Description

Having a selector wait until the state has been initialised would help solve an issue I'm encountering with using NGXS in a Guard, where the initialisation of a particular piece of state is happening too late in the lifecycle.

(Note: this _could_ even be abstracted further to having selectors wait until a particular condition or event has been triggered (e.g. have a selector wait until a particular `Action` has been triggered/completed/etc.))

### Describe the problem you are trying to solve

Given an "auth" state:

```typescript
// `User` model is imported here

export class AuthSetUser {
  static readonly type = '[Auth] Set user';
  constructor(public readonly user: User | null) {}
}

export interface AuthStateModel {
  user: User | null
}

export const AUTH_STATE_TOKEN = new StateToken<AuthStateModel>('auth');

@State<AuthStateModel>({
  name: AUTH_STATE_TOKEN,
  defaults: {
    user: null
  },
})
@Injectable()
export class AuthState implements NgxsOnInit {
  constructor(private authService: AuthService) {}

  ngxsOnInit({ dispatch }: StateContext<AuthStateModel>): void {
    // Async operation here that loads the current user and dispatches the `AuthSetUser` action.
  }

  @Selector([AUTH_STATE_TOKEN])
  static getUser(state: AuthStateModel): User | null {
    return state.user;
  }

  @Selector([AUTH_STATE_TOKEN])
  static isAuthenticated(state: AuthStateModel): boolean {
    return !!state.user;
  }

  @Action(AuthSetUser)
  protected add({ patchState }: StateContext<AuthStateModel>, { user }: AuthSetUser): void {
    patchState({ user });
  }
}
```

… when using this in a guard:

```typescript
canActivate(
  route: ActivatedRouteSnapshot,
  state: RouterStateSnapshot
): Observable<boolean> {
  return this.store.selectOnce(AuthState.isAuthenticated);
}
```

… the guard will receive the value `false`, on initial app load, given the default state contains a `null` user.

I have considered the alternative of having a separate `loaded` field in this auth state (more on this below).

**What would be really handy is having the `getUser` and `isAuthenticated` selectors only become _active_ once the state has initialised (i.e. the Observable stream for each would just not emit anything to start off with).**

### Describe the solution you'd like

Something like:

```typescript

  @Selector([AUTH_STATE_TOKEN], { waitForOnInit: true })
  static getUser(state: AuthStateModel): User | null {
    return state.user;
  }

  @Selector([AUTH_STATE_TOKEN], { waitForOnInit: true })
  static isAuthenticated(state: AuthStateModel): boolean {
    return !!state.user;
  }

```

… which would ensure that the selectors don't emit any values until after the state has been initialised.

### Describe alternatives you've considered

The one alternative I've seen is to have an extra `loaded` field and then access that first before accessing selectors. This creates a dependency on having to remember to inspect this field first, which I think could be flakey and error-prone. Ideally I'd like the public API exposed by the store/state to encapsulate this requirement.

I realise the suggested approach could bring about it's own issues, and is possibly going against state management best practices, so I'm open to any other suggestions.

**Comment by markwhitfeld** [comment]
Hi @jits 
Thank you for your thoughtful suggestion. Unfortunately, this adds too much 'knowledge' to a selector in terms of what it needs to know in order to resolve its value.

A selector should be able to resolve its value at any time you ask. For example, if you were to use `selectSnapshot` instead of `selectOnce` in your code above, what value would it return synchronously? If it returns `undefined` then we are exactly where we started.

Your selectors should be pure functions that do calculations based on their inputs alone and return a result that makes sense.
A selector should not throw errors and you should code defensively as a result.

I recommend the approach that uses the 'isLoading' flag.
If you think about your particular use case, you are essentially saying, I would like the observable that I use for my `canActivate` value to only present a value once the loading has finished.

You could combine two selectors to acheive this:
```TS
  const authenticated$ = this.store.select(AuthState.isLoaded).pipe(
      filter((isLoaded) => isLoaded),
      switchMap(() => this.store.select(AuthState.isAuthenticated))
    );
  }
```

An alternative is to realise that `isAuthenticated` selector could potentially be a tri-state value: true, false, unknown
Let's just say `true`, `false` and `undefined` represents these states. Then your code could be:
```TS
@Selector([AUTH_STATE_TOKEN])
  static isAuthenticated(state: AuthStateModel): boolean {
    return state.isLoaded ? !!state.user : undefined;
  }
```
and
```TS
  const authenticated$ = this.store.select(AuthState.isAuthenticated).pipe(
      filter((isLoaded) => typeof(isLoaded) !== 'undefined')
    );
  }
```
Hope this helps!

**Comment by splincode** [comment]
@markwhitfeld maybe move issue to discussion? for best example

**Comment by jits** [comment]
Hi @markwhitfeld – thanks for the comprehensive response! I agree that this adds too much 'knowledge' to selectors and is very use case specific.

I've done exactly as you suggested – combine the two selectors, in a higher level service that components use. This seems like a reasonable pattern and works well so far. For posterity, here's how I solved it… 

I made a helper function:

```typescript
import { Store } from '@ngxs/store';
import { Observable } from 'rxjs';
import { filter, switchMap } from 'rxjs/operators';

type SelectorFn<T> = (state: any, ...states: any[]) => T;

export const waitUntil = <TResult>(
  store: Store,
  statusSelector: SelectorFn<boolean>,
  resultSelector: SelectorFn<TResult>
): Observable<TResult> =>
  store.select(statusSelector).pipe(
    filter((status) => status),
    switchMap(() => store.select(resultSelector))
  );
```

… which can be used in a service, e.g.:

```typescript
getUser(): Observable<User | null> {
  return waitUntil(this.store, AuthState.isLoaded, AuthState.getUser);
}

isAuthenticated(): Observable<boolean> {
  return waitUntil(this.store, AuthState.isLoaded, AuthState.isAuthenticated);
}
```

I hope this helps anyone else that comes across this problem. And if anyone has ideas on how to improve this further, I'm all ears.

**Comment by jits** [comment]
@splincode –

> @markwhitfeld maybe move issue to discussion? for best example

👍🏽  Nice idea! The examples discussed in this issue could help others.
